docs/design.md

```diff
@@ -38,13 +58,168 @@ spec:
 ```
 
 ### Command
+
 Optionally you can override the default command of a container by providing
 `command` in the Substratus resource:
 ```yaml
 spec:
   command: ["train.sh"]
 ```
 
+### User
+
+Substratus seeks to match the colab environment.
+
+All containers should use the `root` user for all training, importing, and notebook operations. This allows for users to easily install tools with `apt-get` similar to colab. The `/content` directory should correspond to the `WORKDIR` of all Dockerfiles.
+
+NOTE: gvisor can be used to mitigate security risks (See [GKE Sandbox](https://cloud.google.com/kubernetes-engine/docs/concepts/sandbox-pods)).
+
+### Storage
+
+#### Buckets
+
+In Kubernetes the combination of the following fields make up a unique reference to a given object (i.e. database key): `split(.apiVersion, "/")[0] + .kind + .metadata.namespace + .metadata.name`. By storing related artifacts using a similar scheme, restore operations are made trivial: the administrator will not need to worry about backing up the `.status.url` field (as opposed to a scheme where storage location is `.metadata.uid` based - which would be unique for a given in-time instance of an object). When deleting and recreating clusters, if the storage bucket persisted, objects (i.e. Models, Datasets) can simply be re-applied into the new cluster (i.e. [velero](https://velero.io/)/similar is not needed). This plays nicely with GitOps.
+
+Process:
+
+Calculate a hash.
+
+Hashes are used to [prevent bucket hot-spots](https://cloud.google.com/blog/products/gcp/optimizing-your-cloud-storage-performance-google-cloud-performance-atlas).
+
+As opposed to a random id, a hash is used to provide a deterministic pattern storing and finding artifacts. By including a cluster name, Substratus can optionally store artifacts across multiple clusters in a single bucket if a compelling use-case arises in the future.
+
+```sh
+hash_input = "clusters/{cluster}/namespaces/{namespace}/{resource_plural}/{name}"
+
+hash = md5(hash_input)
+```
+
+For example, with the following Model, with a default Substratus installation (cluster = `substratus-1`):
+
+```yaml
+kind: Model
+metadata:
+  name: falcon-7b
+  namespace: team-a
+```
+
+The hash input string would be: `clusters/substratus-1/namespaces/team-a/models/falcon-7b`.
+
+The resulting MD5 hash would be: `f94a0d128bcbd9c1b824e9e5572baf86`.
+
+The following scheme can be used for storing artifacts for Models, Datasets, and Notebooks:
+
+```sh
+# Models
+gs://{bucket}/{hash}/model # Model artifacts (*.pt, etc)
+gs://{bucket}/{hash}/logs  # Logs and converted notebooks
+
+# Datasets
+gs://{bucket}/{hash}/data # Model artifacts (*.pt, etc)
+gs://{bucket}/{hash}/logs # Logs and converted notebooks
+
+# Notebooks
+gs://{bucket}/{hash}/build/{md5-checksum}.tar # Uploaded build context
```

Comment:
I spent a moment thinking about this last night and .zip would be more cross-platform friendly. So far I think we're supporting Windows still

Reply:
oh right the libs would be doing the compression anyway. Yeah tar.gz is it! 👍